waiter/src/waiter/auth/authentication.clj

```diff
@@ -28,21 +29,30 @@
     "Attaches middleware that enables the application to perform authentication.
      The middleware should
      - either issue a 401 challenge asking the client to authenticate itself,
-     - or upon successful authentication populate the request with :authorization/user and :authenticated-principal"))
+     - or upon successful authentication populate the request with :authorization/user and :authorization/principal"))
 
 (defn- add-cached-auth
   [response password principal]
   (cookie-support/add-encoded-cookie response password AUTH-COOKIE-NAME [principal (System/currentTimeMillis)] 1))
 
+(defn assoc-auth-params
+  "Associate values for authenticated user in the request."
+  ([request principal]
+   (assoc-auth-params request (first (str/split principal #"@" 2)) principal))
+  ([request user principal]
+   (assoc request
+          :authorization/principal principal
+          :authorization/user user)))
+
 (defn handle-request-auth
   "Invokes the given request-handler on the given request, adding the necessary
   auth headers on the way in, and the x-waiter-auth cookie on the way out."
-  [request-handler request user principal password]
-  (-> request
-      (assoc :authorization/user user
-             :authenticated-principal principal)
-      request-handler
-      (add-cached-auth password principal)))
+  [handler request user principal password]
+  (let [assoc-auth-params-fn #(assoc-auth-params % user principal)
+        handler' (middleware/wrap-update handler assoc-auth-params-fn)]
```

Comment:
Why would `wrap-context` not work for this scenario? i.e. why did we need the `wrap-update` refactor?

Reply:
`assoc-auth-params` doesn't build a separate map.  It `assoc`es onto an existing map.  We'd either need to create a `auth-params-map` or have a `wrap-update`.  I opted for the latter since `wrap-update` is more flexible than `wrap-merge`.  `wrap-merge` can be trivially implemented by `wrap-update`, but not vice versa.